pyrs/src/pass.rs

```diff
@@ -0,0 +1,147 @@
+use std::{cmp::min, convert::TryInto, fs, num::NonZeroUsize, path::PathBuf};
+
+use pyo3::{prelude::*, types::IntoPyDict};
+use tket2::{json::TKETDecode, op_matches, passes::apply_greedy_commutation, Circuit, T2Op};
+use tket_json_rs::circuit_json::SerialCircuit;
+
+use crate::{
+    circuit::{try_update_hugr, try_with_hugr},
+    optimiser::PyDefaultTasoOptimiser,
+};
+
+#[pyfunction]
+fn greedy_depth_reduce(py_c: PyObject) -> PyResult<(PyObject, u32)> {
+    try_with_hugr(py_c, |mut h| {
+        let n_moves = apply_greedy_commutation(&mut h)?;
+        let py_c = SerialCircuit::encode(&h)?.to_tket1()?;
+        PyResult::Ok((py_c, n_moves))
+    })
+}
+
+/// TASO optimisation pass.
+///
+/// HyperTKET's best attempt at optimising a circuit using circuit rewriting
+/// and TASO.
+///
+/// Input can be in any gate set and will be rebased to Nam, i.e. CX + Rz + H.
+///
+/// Will use at most `max_threads` threads (plus a constant) and take at most
+/// `timeout` seconds (plus a constant). Default to the number of cpus and
+/// 15min respectively.
+///
+/// Log files will be written to the directory `log_dir` if specified.
+///
+/// This requires a `nam_6_3.rwr` file in the current directory. The location
+/// can alternatively be specified using the `rewriter_dir` argument.
+#[pyfunction]
+fn taso_optimise(
+    circ: PyObject,
+    max_threads: Option<NonZeroUsize>,
+    rewriter_dir: Option<PathBuf>,
+    timeout: Option<u64>,
+    log_dir: Option<PathBuf>,
+) -> PyResult<PyObject> {
+    // Runs the following code:
+    // ```python
+    // from pytket.passes.auto_rebase import auto_rebase_pass
+    // from pytket import OpType
+    // auto_rebase_pass({OpType.CX, OpType.Rz, OpType.H}).apply(circ)"
+    // ```
+    Python::with_gil(|py| {
+        let auto_rebase = py
+            .import("pytket.passes.auto_rebase")?
+            .getattr("auto_rebase_pass")?;
+        let optype = py.import("pytket")?.getattr("OpType")?;
+        let locals = [("OpType", &optype)].into_py_dict(py);
+        let op_set = py.eval("{OpType.CX, OpType.Rz, OpType.H}", None, Some(locals))?;
+        let rebase_pass = auto_rebase.call1((op_set,))?.getattr("apply")?;
+        rebase_pass.call1((&circ,)).map(|_| ())
+    })?;
+    nam_taso_optimise(circ, max_threads, rewriter_dir, timeout, log_dir)
+}
+
+/// TASO optimisation pass.
+///
+/// HyperTKET's best attempt at optimising a circuit using circuit rewriting
+/// and TASO.
+///
+/// Input must be in the Nam gate set, i.e. CX + Rz + H.
+///
+/// Will use at most `max_threads` threads (plus a constant) and take at most
+/// `timeout` seconds (plus a constant). Default to the number of cpus and
+/// 30s respectively.
+///
+/// Log files will be written to the directory `log_dir` if specified.
+///
+/// This requires a `nam_6_3.rwr` file in the current directory. The location
+/// can alternatively be specified using the `rewriter_dir` argument.
+#[pyfunction]
+fn nam_taso_optimise(
```

Comment:
This has almost the same doc as `taso_optimise`. The fact that one rebases and the other doesn't should be clear there.

Perhaps just have a `rebase` optional parameter that defaults to `true`?

Reply:
Yes, I've removed the `nam_` variant (I had originially thought I would write some of this in Python, but chose to have it all in Rust in the end).